File: scheduler/driver.go
// Copyright 2023 The Yock Authors. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

// dns, plugin, and driver are all derivatives of the dependency analysis pattern.
// They are now abandoned, see pack/dependency.go for details.
package yocks

import (
	"fmt"
	"path"
	"regexp"

	yocki "github.com/ansurfen/yock/interface"
	yockr "github.com/ansurfen/yock/runtime"
	"github.com/ansurfen/yock/util"

	lua "github.com/yuin/gopher-lua"
)

type yockDriverManager struct {
	plugins yocki.Table
	drivers yocki.Table

	globalDNS *DNS
	localDNS  *DNS
}

func newDriverManager() *yockDriverManager {
	return &yockDriverManager{
		drivers:   yockr.NewTable(),
		plugins:   yockr.NewTable(),
		globalDNS: CreateDNS(util.Pathf("@/global.json")),
		localDNS:  CreateDNS(util.Pathf("@/local.json")),
	}
}

func loadDriver(yocks yocki.YockScheduler) {
	yocks.RegYocksFn(yocki.YocksFuncs{
		"set_driver": driverSetDriver,
		// driver is a callback that works on developer of yock driver
		"driver":      driverDriver,
		"exec_driver": driverExecDriver,
	})
}

// driverSetDriver overloads the implementation of the specified function
//
// @param name string
//
// @param fn function
func driverSetDriver(yocks yocki.YockScheduler, state yocki.YockState) int {
	ys := yocks.(*YockScheduler)
	ys.getDrivers().Value().RawSetString(state.CheckString(1), state.CheckFunction(2))
	return 0
}

/*
* @param driver string
* @param name string
* @return string
 */
func driverDriver(yocks yocki.YockScheduler, state yocki.YockState) int {
	driver := state.CheckString(1)
	name := state.CheckString(2)
	out, err := util.ReadStraemFromFile(path.Join(util.DriverPath, name+".lua"))
	if err != nil {
		panic(err)
	}
	reg := regexp.MustCompile(`driver\s*\((.*)function`)
	did := driver + "_" + name
	yocks.Eval(reg.ReplaceAllString(string(out), fmt.Sprintf(`driver("%s",function`, did)))
	yocks.SetGlobalVar(driver, yocks.(*YockScheduler).getDrivers().Value().RawGetString(did))
	state.Push(lua.LString(did))
	return 1
}

// @param name string
//
// @param args ...string
func driverExecDriver(yocks yocki.YockScheduler, state yocki.YockState) int {
	if lv := yocks.(*YockScheduler).getDrivers().Value().RawGetString(state.CheckString(1)); lv.Type() == lua.LTFunction {
		args := []lua.LValue{}
		for i := 3; i <= state.Argc(); i++ {
			args = append(args, state.CheckLValue(i))
		}
		yocks.EvalFunc(lv.(*lua.LFunction), append([]lua.LValue{state.CheckLTable(2)}, args...))
	}
	return 0
}

File: lib/deprecated/doc/driver.lua
-- Copyright 2023 The Yock Authors. All rights reserved.
-- Use of this source code is governed by a MIT-style
-- license that can be found in the LICENSE file.

-- dns, plugin, and driver are all derivatives of the dependency analysis pattern.
-- They are now abandoned, see pack/dependency.go for details.

---@meta _

---@param tbl table
function pull(tbl) end

---@param plugin string
---@return string,string
function parse_plugin(plugin) end

---@param file string
---@return string
function export_builder(file) end

---@class pluginlist
plugin_list = {}

---@param path string
---@return boolean
function plugin_list:IsExist(path) end

---@param pid string
---@param path string
function plugin_list:AddPlugin(pid, path) end

---@class exportOpt
---@field update fun()
---@field install fun()
---@field uninstall fun()
---@field init fun(env: any)
local exportOpt = {}

---@param opt exportOpt
function export(opt) end

---@param opt table
---@vararg string
function installs(opt, ...) end

---@param plugin string
---@param opt table
function install(plugin, opt) end

---@param file string
---@return string
function load_plugin(file) end

plugins = {}

---@param opt table
function plugin(opt) end

---@class ldns
ldns = {}

---@param domain string
function ldns:GetDriver(domain) end

---@param domain string
function ldns:GetPlugin(domain) end

---@param domain string
---@param url string
---@param path string
function ldns:PutPlugin(domain, url, path) end

---@param domain string
---@param url string
---@param path string
function ldns:PutDriver(domain, url, path) end

---@param domain string
---@param alias string
function ldns:AliasDriver(domain, alias) end

---@param domain string
---@param alias string
function ldns:AliasPlugin(domain, alias) end

---@class gdns
gdns = {}

---@param domain string
function gdns:GetDriver(domain) end

---@param domain string
---@return table
function gdns:GetPlugin(domain) end

---@param domain string
---@param url string
---@param path string
function gdns:UpdatePlugin(domain, url, path) end

---@param callback fun(...): ...:any
function driver(callback) end

---@param fn string
---@return fun(opt: table, ...:any)
function uninit_driver(fn)
    ---@param opt table
    ---@vararg any
    return function(opt, ...) end
end

---@param driver string
---@param name string
---@return string
function set_driver(driver, name) end

---@param did string
---@param opt table
---@vararg any
function exec_driver(did, opt, ...) end
